hack/generated/controllers/reflect_helpers.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"github.com/Azure/k8s-infra/hack/generated/pkg/genruntime"
+	"github.com/Azure/k8s-infra/hack/generated/pkg/util/armresourceresolver"
+	"github.com/pkg/errors"
+	"reflect"
+)
+
+// ResourceSpecToArmResourceSpec converts a genruntime.MetaObject (a Kubernetes representation of a resource) into
+// a genruntime.ArmResourceSpec - a specification which can be submitted to Azure for deployment
+func ResourceSpecToArmResourceSpec(
+	ctx context.Context,
+	resolver *armresourceresolver.Resolver,
+	metaObject genruntime.MetaObject) (string, genruntime.ArmResourceSpec, error) {
+
+	metaObjReflector := reflect.Indirect(reflect.ValueOf(metaObject))
+	if !metaObjReflector.IsValid() {
+		return "", nil, errors.Errorf("couldn't indirect %T", metaObject)
+	}
+
+	specField := metaObjReflector.FieldByName("Spec")
+	if !specField.IsValid() {
+		return "", nil, errors.Errorf("couldn't find spec field on type %T", metaObject)
+	}
+
+	// Spec fields are values, we want a ptr
+	specFieldPtr := reflect.New(specField.Type())
+	specFieldPtr.Elem().Set(specField)
+
+	// TODO: how to check that this doesn't fail
+	spec := specFieldPtr.Interface()
+
+	armTransformer, ok := spec.(genruntime.ArmTransformer)
+	if !ok {
+		return "", nil, errors.Errorf("spec was of type %T which doesn't implement genruntime.ArmTransformer", spec)
+	}
+
+	resourceGroupName, azureName, err := resolver.GetFullAzureNameAndResourceGroup(ctx, metaObject)
+	if err != nil {
+		return "", nil, err
+	}
+
+	armSpec, err := armTransformer.ConvertToArm(azureName)
+	if err != nil {
+		return "", nil, errors.Wrapf(err, "failed to transform resource %s to ARM", metaObject.GetName())
+	}
+
+	typedArmSpec, ok := armSpec.(genruntime.ArmResourceSpec)
+	if !ok {
+		return "", nil, errors.Errorf("failed to cast armSpec of type %T to genruntime.ArmResourceSpec", armSpec)
+	}
+
+	return resourceGroupName, typedArmSpec, nil
```

Comment:
RG is an `owner`, so it's part of the spec in that sense - but it's not a field called `resourceGroup`, it's just the topmost owner.

It's separate here because it actually needs to be separate for the deployment we're going to do.

If I have say a vnet myvnet with a subnet mysubnet, in RG myrg, in Kube thats:
mysubnet owner: myvent
myvnet owner: myrg
myrg (no owner)

translated into Azure terminology, that's created by issuing a deployment for
myvnet/mysubnet
to the resource group "myrg"

Reply:
With all of this said - I agree that this function (3 returns, RG is special) is a bit of a smell.

We could maybe do something like:
1. `FindOwnerHierarchy`: `genruntime.MetaObject -> []genruntime.MetaObject`
2. `HierarchyToResourceGroupAndAzureName`: `[]genruntime.MetaObject -> (string, string)`

This splits it a bit cleaner into two functions each with a single return and I think that second function wouldn't need an error return as it couldn't fail?

Would that be cleaner?

Reply:
(Note that RG is still special in the proposed alternative, I'm not sure how to avoid that at this level as it's special as far as Azure is concerned too)

Reply:
Oh yeah, that's heaps better!

Reply:
let me split this off into another PR - I have a good commit it can go with where I refactor the caller of this function some too.